### PR TITLE
dispose graphics in TabControl.WmReflectDrawItem()

### LINKED ISSUE #10534: Memory leak in TabControlWmReflectDrawItem(), dispose not called on graphics
### .NET version

8.0

### Did it work in .NET Framework?

Not tested/verified

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

unknown

### Issue description

See code below.  The `Graphics` object created by `dis->hDC.CreateGraphics()` is not disposed (neither via `DrawItemEventArgs.Dispose()` nor `WmReflectDrawItem()`).  This results in the object having its finalizer run.

https://github.com/dotnet/winforms/blob/9e2fc95fca4e6d18444c819b35bb1338ac6ca63d/src/System.Windows.Forms/src/System/Windows/Forms/Controls/TabControl/TabControl.cs#L1935

```
    private unsafe void WmReflectDrawItem(ref Message m)
    {
        DRAWITEMSTRUCT* dis = (DRAWITEMSTRUCT*)(nint)m.LParamInternal;

        using DrawItemEventArgs e = new(
            dis->hDC.CreateGraphics(),
            Font,
            dis->rcItem,
            (int)dis->itemID,
            (DrawItemState)(int)dis->itemState);

        OnDrawItem(e);

        m.ResultInternal = (LRESULT)1;
    }
```

### Steps to reproduce

n/a

### PR DESCRIPTION
**Pull request by jgilbert2017**
fixes #10534

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/10536)

### CODE REVIEW

**Review by Tanya-Solyanik — APPROVED** [review]

**Comment by jgilbert2017** [comment]
@dotnet-policy-service agree

**Review by Tanya-Solyanik — APPROVED** [review]
There ae more instances of CreateGraphicsInternal that could be fixed. Would you be able to take care of them as well?

**Comment by jgilbert2017** [comment]
@Tanya-Solyanik sure.  can you point them out?  i checked ListControl and ComboBox that @kirsan31 mentioned and the bad code doesn't exist on the main branch.

**Comment by Tanya-Solyanik** [comment]
List box is correct in a sense that it has a try/finally, but a `using` would simplify the code.

ComponentEditorForm, ThreadExceptionDialog, SplitterContainer do not have try/finally.

**Comment by Tanya-Solyanik** [comment]
@jgilbert2017  - what version of .NET do you need this fixed in? Does this affect your production code?

**Comment by jgilbert2017** [comment]
@Tanya-Solyanik a net8 patch would be ideal :-).  i've gotten sporadic production OOM exceptions due to the tab control lack of Dispose() call.

**Comment by jgilbert2017** [comment]
@Tanya-Solyanik lmk know what you think i should do.  i'd be happy to create another PR to do a more thorough cleanup.

**Comment by Tanya-Solyanik** [comment]
@jgilbert2017  - 
Would you be able to confirm that this change fixes your sporadic OOM exceptions? Run your app on a private version built off release\8.0 in test environment perhaps? I would have expected your application to freeze for the time it takes the finalizer to run, due to this bug.

I would prefer:
1. a very targeted PR with the fix (i.e. your first version) and a unit test against Main. 
2. PR with the same change against release/8.0
3. a follow up PR against main that enforces the `using` pattern for other instances of CreateGraphics* methods

**Comment by kirsan31** [comment]
> i've gotten sporadic production OOM exceptions due to the tab control lack of Dispose() call.

As I mentioned [here](https://github.com/dotnet/winforms/issues/10534#issuecomment-1870148689) you should not have OOM (or other leaks problems) due to `Graphics.Dispose()` not being called. Because `Graphics` have a finalizer. Unnecessary finalization must be only a performance problem. Unless of course we have an error in the garbage collector logic 🙄

**Comment by jgilbert2017** [comment]
@kirsan31 graphics handles are limited resources and finalization may be arbitrarily delayed causing a bunch of objects to be pending finalization and sometimes i like to switch tabs very quickly ;-).

**Comment by jgilbert2017** [comment]
@kirsan31 just to be clear, the OOME is not due to being out of memory, it's due to failure to allocate a new handle.

**Comment by kirsan31** [comment]
> @kirsan31 just to be clear, the OOME is not due to being out of memory, it's due to failure to allocate a new handle.

So you are reaching GDI/USER objects limit - make sense... To be sure that we caught the root - this problem can (mast) be solved with manual triggering GC?

**Comment by kirsan31** [comment]
Another thing that worries me is that in the [original version](https://github.com/dotnet/winforms/commit/e9fc1dac4f9792856f0bc8932486303dd8a8fe02#diff-edd2fb85653c19f4c380ec1ae98b35d6723ca3ecb80ef29a4be708f83fa30463L1949) we pass HDC - not Graphics. For `ComboBox` and `ListBox` this [was fixed](https://github.com/dotnet/winforms/issues/10534#issuecomment-1870591253) later. Bat not for `TabControl`.
Passing HDC will result:
https://github.com/dotnet/winforms/blob/d4caca7a596df9e6d2c4854fb6f14f284ec6ad0c/src/System.Windows.Forms/src/System/Windows/Forms/Rendering/DrawingEventArgs.cs#L99-L113
And `CreateGraphics`:
https://github.com/dotnet/winforms/blob/d4caca7a596df9e6d2c4854fb6f14f284ec6ad0c/src/System.Drawing.Common/src/System/Drawing/Graphics.cs#L172-L176

I think the code is a little different 🤔 Could this change lead to undesirable consequences, for example #10462?
In the end, I would [agree with @elachlan](https://github.com/dotnet/winforms/issues/10534#issuecomment-1870644909) and return the original version with HDC...

**Comment by Tanya-Solyanik** [comment]
@kirsan31 @elachlan - Original version was using graphics - https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/TabControl.cs,5780e62ffb386499
[PR9309](https://github.com/dotnet/winforms/commit/e9fc1dac4f9792856f0bc8932486303dd8a8fe02#diff-edd2fb85653c19f4c380ec1ae98b35d6723ca3ecb80ef29a4be708f83fa30463) was undoing what [PR3553](https://github.com/dotnet/winforms/commit/a6a308fa452770caafca55ee4d8ac54d1ee1015d#diff-edd2fb85653c19f4c380ec1ae98b35d6723ca3ecb80ef29a4be708f83fa30463) had introduced. Keeping graphics seems correct.

**Comment by kirsan31** [comment]
@Tanya-Solyanik 
> [PR9309](https://github.com/dotnet/winforms/commit/e9fc1dac4f9792856f0bc8932486303dd8a8fe02#diff-edd2fb85653c19f4c380ec1ae98b35d6723ca3ecb80ef29a4be708f83fa30463) was undoing what [PR3553](https://github.com/dotnet/winforms/commit/a6a308fa452770caafca55ee4d8ac54d1ee1015d#diff-edd2fb85653c19f4c380ec1ae98b35d6723ca3ecb80ef29a4be708f83fa30463) had introduced. Keeping graphics seems correct.

Then we need to restore `CreateGraphic` for `ComboBox` and `ListBox`. Before PR3553 all of them (`ComboBox`, `ListBox`, `TabControl`) were using `CreateGraphic`.

**Comment by jgilbert2017** [comment]
The comment below is on the ctor that takes the HDC.  The graphics object is ultimately passed to `OnDrawItem` which can be implemented by the user and subject to breaking the assumption below.  This would seem to imply the ctor that takes Graphics is the one to use.

```
    /// <summary>
    ///  For internal use to improve performance. DO NOT use this method if you modify the Graphics Clip or Transform.
    /// </summary>
    internal Graphics GetOrCreateGraphicsInternal(Action<Graphics>? creationAction = null)
```

**Comment by elachlan** [comment]
@JeremyKuhne I think can comment on if we should be passing the device context or graphics here.

Jeremy did the work refactoring away from graphics to device context (HDC) to avoid creating graphics unnecessarily to improve performance. This is to do with GDI being more performant than GDI+

**Comment by kirsan31** [comment]
> This would seem to imply the ctor that takes Graphics is the one to use.

Yea we discussed this with @Tanya-Solyanik above - originally there were `CreateGraphic`. Then I suggested revert `ComboBox` and `ListBox` to `CreateGraphic` too.

**Comment by Tanya-Solyanik** [comment]
@kirsan31  

>Then I suggested revert ComboBox and ListBox to CreateGraphic too.

There is an issue of pallet to be double-checked. 

Original code was setting up a halftone pallet on the DC before creating the Graphics object.
EventArg constructor that takes the DC is still doing that - [here](https://github.com/dotnet/winforms/blob/58ffd1a410d00f7f5b6e609ddf88114117c0e1c0/src/System.Windows.Forms/src/System/Windows/Forms/Rendering/DrawingEventArgs.cs#L93). 
CreateGraphics route is not setting up this pallet. So the question is to @JeremyKuhne  - why did he move back to CreateGraphics here - [PR9309](https://github.com/dotnet/winforms/commit/e9fc1dac4f9792856f0bc8932486303dd8a8fe02#diff-edd2fb85653c19f4c380ec1ae98b35d6723ca3ecb80ef29a4be708f83fa30463)

**Comment by JeremyKuhne** [comment]
Sorry, I've been out.

The simple answer is that I make mistakes. :) Intellisense gave me significant grief when I was converting this so I'm not completely surprised that I messed this up.

Creating the `DrawingEventArgs` via `HDC` is _always_ preferable if you haven't created the `Graphics` object first. `IGraphicsHdcProvider` is then able to unwrap the `HDC` when drawing via GDI internally without creating the `Graphics` object.

File: src/System.Windows.Forms/src/System/Windows/Forms/Controls/ComboBox/ComboBox.cs [hunk 1]
```diff
@@ -3566,7 +3566,7 @@ private void UpdateItemHeight()
         else if (DrawMode == DrawMode.OwnerDrawVariable)
         {
             PInvoke.SendMessage(this, PInvoke.CB_SETITEMHEIGHT, (WPARAM)(-1), (LPARAM)ItemHeight);
-            Graphics graphics = CreateGraphicsInternal();
+            using Graphics graphics = CreateGraphicsInternal();
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/Controls/ComboBox/ComboBox.cs:3569]: I don't know how often consuming code never uses the `Graphics`. Possible performance optimization in the future to make this lazy in `MeasureItemEventArgs` and make the args disposable.

File: src/System.Windows.Forms/src/System/Windows/Forms/Controls/TabControl/TabControl.cs [hunk 2]
```diff
@@ -1936,8 +1936,9 @@ private unsafe void WmReflectDrawItem(ref Message m)
     {
         DRAWITEMSTRUCT* dis = (DRAWITEMSTRUCT*)(nint)m.LParamInternal;
 
+        using var graphics = dis->hDC.CreateGraphics();
         using DrawItemEventArgs e = new(
-            dis->hDC.CreateGraphics(),
+            graphics,
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/Controls/TabControl/TabControl.cs:1941]: This really should use the `dis->hDC` in the overload that takes `HDC`.

**Comment by elachlan** [comment]
@jgilbert2017 Did you need anymore guidance on this issue?

**Comment by jgilbert2017** [comment]
@Tanya-Solyanik to recap... i can resubmit a patch
- which branch do you want it against?
- i will use the dis->hDC ctor() per @JeremyKuhne 's comment

**Comment by elachlan** [comment]
@jgilbert2017 against main branch please.

**Comment by Tanya-Solyanik** [comment]
>i will use the dis->hDC ctor() per @JeremyKuhne 's comment

yes, but there will be several places that regressed in [9309](https://github.com/dotnet/winforms/pull/9309)

>@jgilbert2017 against main branch please.

yes

**Comment by jgilbert2017** [comment]
@Tanya-Solyanik can you point them out? I'm only aware of the TabControl issue. i checked ListControl and ComboBox that @kirsan31 mentioned and the bad code doesn't exist on the main branch.

**Comment by kirsan31** [comment]
@jgilbert2017
If I understand right (we need hDC everywhere) so only TabControl need to fixed.

**Comment by Tanya-Solyanik** [comment]
>@Tanya-Solyanik can you point them out? 

I only looked at the PR, yes, combobox and listbox are fixed in main

**Review by Tanya-Solyanik — APPROVED** [review]

**Pull request merged**